File: matlab/one-shot-classification/demo_classification.m
% Compute one-shot classification error for
% Modified Hausdorff Distance baseline.
%
% Running this code should lead to a result of 38.8 percent errors.
%
%   M.-P. Dubuisson, A. K. Jain (1994). A modified hausdorff distance for object matching.
%     International Conference on Pattern Recognition, pp. 566-568.
% 
% You can plug in your own fucntion for the variable 'fcost',
%  and set ftype='cost' to  ftype='score' depending on whether a low or high value is
%  good.
%
% ** Models must be trained on data_background.mat to avoid 
% using images and alphabets used in the one-shot evaluation **
%
function demo_classification()

    addpath(genpath('..'));
    
    load('items_classification','nrun','cell_train','cell_test','cell_Y');
    %   20 different mini one-shot classification experiments,
    %   where each cell is a different experiment

    fcost = @(itrain,itest) HD(itrain,itest);
    
    fprintf(1,'One-Shot Classification (Modified Hausdorff Distance)\n');
    perror = zeros(nrun,1);
    for r=1:nrun
        trainset = cell_train{r};
        testset = cell_test{r};
        Y = cell_Y{r};
        perror(r) = myclassify(trainset,testset,fcost,Y,'cost');
        fprintf(1,' run %d (error %s%%)\n',r,num2str(perror(r),3));
    end
    fprintf(1,'average error: %s%%\n',num2str(mean(perror),3));
        
end

function cost = HD(itrain,itest)
    % Return distance (cost) between training (itrain) and test image (itest)
    assert(UtilImage.check_black_is_true(itrain.img));
    assert(UtilImage.check_black_is_true(itest.img));
    [irow,icol] = find(itrain.img);
    dtrain = [irow icol];
    dtrain = normalize(dtrain);
    [irow,icol] = find(itest.img);
    dtest = [irow,icol];
    dtest = normalize(dtest);
    cost = ModHausdorffDist(dtrain,dtest);
end


function d = normalize(d)
    % Subtract off the center of mass for set of points
    % rows of d are points
    assert(size(d,2)==2);
    d1 = d(:,1);
    d2 = d(:,2);
    d(:,1) = d(:,1) - mean(d1);
    d(:,2) = d(:,2) - mean(d2);
end

function mhd = ModHausdorffDist(A,B)
    % Input
    %  A : [n x 2] x,y coordinates of "on" pixels
    %  B : [m x 2] x,y coordinates of "on" pixels
    % Output
    %  mhd: distance between images
    assert(size(A,2)==2);
    assert(size(B,2)==2);
    D = pdist2(A,B);
    mindist_A = min(D,[],2);
    mindist_B = min(D,[],1);
    mean_A = mean(mindist_A);
    mean_B = mean(mindist_B);
    mhd = max(mean_A,mean_B);
end

File: python/one-shot-classification/demo_classification.py
import numpy as np
import copy
from scipy.ndimage import imread
from scipy.spatial.distance import cdist

# Parameters
nrun = 20 # number of classification runs
fname_label = 'class_labels.txt' # where class labels are stored for each run

def classification_run(folder,f_load,f_cost,ftype='cost'):
	# Compute error rate for one run of one-shot classification
	#
	# Input
	#  folder : contains images for a run of one-shot classification
	#  f_load : itemA = f_load('file.png') should read in the image file and process it
	#  f_cost : f_cost(itemA,itemB) should compute similarity between two images, using output of f_load
	#  ftype  : 'cost' if small values from f_cost mean more similar, or 'score' if large values are more similar
	#
	# Output
	#  perror : percent errors (0 to 100% error)
	# 
	assert ((ftype=='cost') | (ftype=='score'))

	# get file names
	with open(folder+'/'+fname_label) as f:
		content = f.read().splitlines()
	pairs = [line.split() for line in content]
	test_files  = [pair[0] for pair in pairs]
	train_files = [pair[1] for pair in pairs]
	answers_files = copy.copy(train_files)
	test_files.sort()
	train_files.sort()	
	ntrain = len(train_files)
	ntest = len(test_files)

	# load the images (and, if needed, extract features)
	train_items = [f_load(f) for f in train_files]
	test_items  = [f_load(f) for f in test_files ]

	# compute cost matrix
	costM = np.zeros((ntest,ntrain),float)
	for i in range(ntest):
		for c in range(ntrain):
			costM[i,c] = f_cost(test_items[i],train_items[c])
	if ftype == 'cost':
		YHAT = np.argmin(costM,axis=1)
	elif ftype == 'score':
		YHAT = np.argmax(costM,axis=1)
	else:
		assert False

	# compute the error rate
	correct = 0.0
	for i in range(ntest):
		if train_files[YHAT[i]] == answers_files[i]:
			correct += 1.0
	pcorrect = 100 * correct / ntest
	perror = 100 - pcorrect
	return perror

def ModHausdorffDistance(itemA,itemB):
	# Modified Hausdorff Distance
	#
	# Input
	#  itemA : [n x 2] coordinates of "inked" pixels
	#  itemB : [m x 2] coordinates of "inked" pixels
	#
	#  M.-P. Dubuisson, A. K. Jain (1994). A modified hausdorff distance for object matching.
	#  International Conference on Pattern Recognition, pp. 566-568.
	#
	D = cdist(itemA,itemB)
	mindist_A = D.min(axis=1)
	mindist_B = D.min(axis=0)
	mean_A = np.mean(mindist_A)
	mean_B = np.mean(mindist_B)
	return max(mean_A,mean_B)

def LoadImgAsPoints(fn):
	# Load image file and return coordinates of 'inked' pixels in the binary image
	# 
	# Output:
	#  D : [n x 2] rows are coordinates
	I = imread(fn,flatten=True)
	I = np.array(I,dtype=bool)
	I = np.logical_not(I)
	(row,col) = I.nonzero()
	D = np.array([row,col])
	D = np.transpose(D)
	D = D.astype(float)
	n = D.shape[0]
	mean = np.mean(D,axis=0)
	for i in range(n):
		D[i,:] = D[i,:] - mean
	return D

if __name__ == "__main__":
	#
	# Running this demo should lead to a result of 38.8 percent errors.
	#
	#   M.-P. Dubuisson, A. K. Jain (1994). A modified hausdorff distance for object matching.
	#     International Conference on Pattern Recognition, pp. 566-568.
	#
	# ** Models should be trained on images in 'images_background' directory to avoid 
	#  using images and alphabets used in the one-shot evaluation **
	#
	print 'One-shot classification demo with Modified Hausdorff Distance'
	perror = np.zeros(nrun)
	for r in range(1,nrun+1):
		rs = str(r)
		if len(rs)==1:
			rs = '0' + rs		
		perror[r-1] = classification_run('run'+rs, LoadImgAsPoints, ModHausdorffDistance, 'cost')
		print " run " + str(r) + " (error " + str(	perror[r-1] ) + "%)"		
	total = np.mean(perror)
	print " average error " + str(total) + "%"